Fix #413 parse entities positioning

unintended opened the pull request:
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context.

Fixes #413

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

JrooTJunior commented:
LGTM

Can you also make backport to `dev-3.x` branch with adding corresponding test that prove you changes?


And that is something strange. When i check entity length after adding MarkdownV2 i didn't see the need to use `utf-16-le` and then i remove it. So as i an see in your description i was wrong... Thanks

JrooTJunior commented on aiogram/utils/text_decorations.py:
```diff
@@ -57,14 +57,14 @@ def unparse(self, text: str, entities: Optional[List[MessageEntity]] = None) ->
         result = "".join(
             self._unparse_entities(
-                text, sorted(entities, key=lambda item: item.offset) if entities else []
+                text.encode('utf-16-le'), sorted(entities, key=lambda item: item.offset) if entities else []
```
Small optimization.

Please move calling of encode to separated method:
```python3
def _add_surrogates(self, text: str) -> bytes:
    return text.encode('utf-16-le')
``` 
And also use adding/removing surrogates inside `_unparse_entities` method

JrooTJunior commented on aiogram/utils/text_decorations.py:
```diff
@@ -57,14 +57,14 @@ def unparse(self, text: str, entities: Optional[List[MessageEntity]] = None) ->
     def _unparse_entities(
         self,
-        text: str,
+        text: bytes,
```
I think text should be passed to this function without any changes and then encoding should be placed inside this method

JrooTJunior commented on aiogram/utils/text_decorations.py:
```diff
@@ -74,15 +74,15 @@ def _unparse_entities(
+            if entity.offset * 2> offset:
+                yield self.quote(text[offset : entity.offset * 2].decode('utf-16-le'))
+            start = entity.offset * 2
+            offset = entity.offset * 2 + entity.length * 2
```
Very small optimization:
```suggestion
            offset = (entity.offset + entity.length) * 2
```

Or maybe calculate doubled size once and then use it

For example:
```python3
for index, entity in enumerate(entities):
    entity_length = entity.length * 2
    entity_offset = entity.offset * 2


    if entity_offset < offset:
        continue
    ...
```

unintended replied:
I was trying to do that way, but there is recursive call of _unparse_entities, so it may lead to unexpected results for nested entities. I didn't have time to check the behavior and tine this index magic, so I decided to move decode/encode outside and pass the bytes. That's why I called the solution "raw" :)

JrooTJunior replied:
Hm.. Just try it and if this way is hard leave it as is 🙂

JrooTJunior replied:
And with recursive call, hmm, encoding outside `_unparse_entities` make sense..

JrooTJunior commented on aiogram/utils/text_decorations.py:
```diff
@@ -74,15 +74,15 @@ def _unparse_entities(
-            start = entity.offset
-            offset = entity.offset + entity.length
+            if entity.offset * 2> offset:
+                yield self.quote(text[offset : entity.offset * 2].decode('utf-16-le'))
```
Move decoding to separated method:

```python3
def _remove_surrogates(self, data: bytes) -> str:
    return data.decode('utf-16-le')
```

And then use it here as
```python3
yield self.quote(self._remove_surrogates(text[offset : entity_offset]))
```

JrooTJunior commented on aiogram/utils/text_decorations.py:
```diff
@@ -94,7 +94,7 @@ def _unparse_entities(
 
         if offset < length:
-            yield self.quote(text[offset:length])
+            yield self.quote(text[offset:length].decode('utf-16-le'))
```
```python3
yield self.quote(self._remove_surrogates(text[offset:length]))
```

unintended commented:
I updated the code: added tests, refactored encoding to methods, but didn't manage to put surrogates methods inside `_unparse_entities`. Maybe someone will try to make make it one day :) 
Also tried to port to dev-3.x, but with no luck yet (couldn't port the tests properly becoase of different nature of api classes) and I don't have much time to do it, I'm sorry.

JrooTJunior commented:
in `dev-3.x` you can add test-case to parameterized test of decorations:
https://github.com/aiogram/aiogram/blob/6f53f15577e3b05a0268b78efdfb410e95672429/tests/test_utils/test_text_decorations.py#L110-L231 

and at the end of list add item:
```python3
[
    markdown_decoration,
    "🚀 i'm bold and italic and still bold",
    [
        MessageEntity(offset=3, length=34, type="bold"),
        MessageEntity(offset=12, length=10, type="italic"),
    ],
    "🚀 *i'm bold _and italic_\r and still bold*"
]
```

JrooTJunior commented:
Thanks for contribution :wink:
